wordpress-mobile/WordPress-Android
# Handle plugin change events in PluginBrowserViewModel

Please note that before reviewing this PR, both #7280 & https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/722 needs to be reviewed.

**Using background threads**

We want `PluginBrowserViewModel` to work in a bg thread as much as possible. I've introduced `@WorkerThread` annotation for this to a lot of methods. I feel like we should mark the whole class with it, but I am not sure if any future change can be risky that way. As a result of this change, and probably a good practice in general, now the LiveData models are updated with `postValue` instead of `setValue` which ensures the update happens in the main thread. Finally, the FluxC events are marked with `BACKGROUND` thread instead of the `MAIN` one since we don't directly update the UI from any of it. These changes will hopefully make things much more responsive for the user.

**Properly handling FluxC events in `PluginBrowserViewModel`**

Due to the changes in #7280 and some oversight by me in the initial implementation of the `PluginBrowserViewModel` not all plugin change events were being handled. When any of these events happen, we can simply re-query all the plugins from store, but that can become very expensive especially if a lot of wporg plugins are fetched quickly.

To avoid constant UI updates, I decided to go with the following optimizations:

1. All single plugin update events share a set, whenever a successful change happens, we'll add the slug of the plugin to this set and trigger an update to all lists if necessary.
2. There is a 250ms delay to combine these events. It could be that multiple plugins are updated, or a single plugin had multiple updates (activated and enabled autoupdates etc). Updating the UI is an expensive operation and we currently do it with `notifyDataSetChanged` which is not recommended. The less time we have to do this, the better it will be.
3. Once we decide that we have items to update, we prepare a map that has the slug as the key and plugin as the value. For each list (except for site plugins), we check whether the list item's slug is in the key set, if so we replace the item, if not keep the old item. We do this with a map instead of 1 by 1, so that we can post a single event to the UI that it's list is updated.
4. Site plugins is a special case, because plugins might be installed or removed, it can be really tricky to keep that list up to date. Although possible, I don't think it's worth the effort and might even be more costly than asking the store for the up to date information. Also, we care more about site plugins than anything else.

I added a lot of comments to point out some issues, so hopefully that'll be enough. If anything is not very clear let me know and I'll add comment for that as well!

To test:
* Installing a fresh copy of the app to test how fast the plugin page is loading and if there is any input lag would be a good test for this PR
* Aside from that, all plugin actions should be tested and it's affects should be checked in the plugin lists.

I've recorded a couple videos to showcase this PR a bit. I unfortunately forgot about the search list in my first video, so the second one is only about that. https://cloudup.com/czbFw8tBsIG & https://cloudup.com/c8FmkJjOfXt. Please note that it was actually impossible to update the search results before this change since it's not kept in the store. However, as it can be seen from the second video, it works pretty well here.

/cc @theck13 (for the review) @nbradbury (to keep you in the loop for plugin changes)
